src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildTriggerListener.java

```diff
@@ -106,13 +106,8 @@ private void updateDownstreamBuildAction(Run<?, ?> run) {
                             LOGGER.log(Level.FINE, () -> "Unable to update DownstreamBuildAction for " + upstream + " node " + buildUpstreamCause.getNodeId());
                             continue;
                         }
-                        DownstreamBuildAction downstreamAction = node.getPersistentAction(DownstreamBuildAction.class);
-                        if (downstreamAction == null) {
-                            // Should only happen for builds already in the queue when this plugin is updated to include DownstreamBuildAction.
-                            downstreamAction = new DownstreamBuildAction(run.getParent());
-                            node.addAction(downstreamAction);
-                        }
-                        downstreamAction.setBuild(run);
+                        DownstreamBuildAction action = new DownstreamBuildAction(run);
+                        node.addOrReplaceAction(action);
                         run.save();
```

Comment:
Is the `run.save()` still needed?

Reply:
`WorkflowRun.save` + `CpsFlowExecution$ConverterImpl` will not flush the `FlowNodeStorage` as far as I can tell, so the `run.save` call seems pointless.

The situation seems complicated though. Mutation of the list returned by `FlowNode.getActions` calls `FlowNode.save` which calls `FlowActionStorage.saveActions`, and `CpsFlowExecution` routes that to `FlowNodeStorage.saveActions`. For `SimpleXStreamFlowNodeStorage` in most cases I think this will actually write the changes to disk, but not always. For `BulkFlowNodeStorage` it will just set a flag that something has been modified.

`CpsFlowExecution.saveOwner` and `FlowNodeStorage.storeNode(node, false)` would flush the `FlowNodeStorage`, but are not accessible here. I think that with the current code, if you use `build(wait: false, ...)`, and the upstream build completes before the downstream build starts, and you are running https://github.com/jenkinsci/workflow-cps-plugin/pull/807 or using `PERFORMANCE_OPTIMIZED`, then nothing will ever save the change to disk.

Not sure if there is a simple fix. We could maybe switch to storing this action along with the `FlowNode` IDs on `Run` instead. I guess this general issue applies to any attempts to modify actions on a `FlowNode` for a possibly-completed Pipeline run (probably very uncommon). We could introduce a new overload of `FlowActionStorage.saveActions` that forces the storage to be flushed if we care about the general case.

Reply:
> We could maybe switch to storing this action along with the `FlowNode` IDs on `Run` instead.

See https://github.com/jenkinsci/pipeline-build-step-plugin/pull/13#issuecomment-1847847813. https://github.com/search?q=org%3Ajenkinsci%20DownstreamBuildAction&type=code shows no usages yet so it would be possible to do so without breaking anything. That may be better than trying to fix

> any attempts to modify actions on a `FlowNode` for a possibly-completed Pipeline run

though if we believe that to be unsupported (compare `onFinalized` w.r.t. the log file) then it would be better to explicitly throw an `IOException` saying so. (Whereas there are some established reasons why `<actions>` in `build.xml` might be modified after completion.)

Either way, it seems that

> the upstream build completes before the downstream build starts

deserves some test coverage too.

Reply:
Yes I think I will switch to storing the data on the `Run`.

> though if we believe that to be unsupported (compare onFinalized w.r.t. the log file) then it would be better to explicitly throw an IOException saying so. 

Yeah I agree.

> Either way, it seems that
> > the upstream build completes before the downstream build starts
>
> deserves some test coverage too.

The current test does cover this case, it's just that we haven't picked up https://github.com/jenkinsci/workflow-cps-plugin/pull/807 yet and aren't using `PERFORMANCE_OPTIMIZED`, so it passes. I will try to expand the coverage though.

Reply:
See https://github.com/jenkinsci/pipeline-build-step-plugin/pull/128/commits/cd63d16e2d6a3b5aa4736c73ff1e2a87b1e7b44a.